---
title: "burden_analyses_lmm_publication_clean"
author: "Sigurgeir Olafsson"
date: "11/3/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Burden of somatic mutation in inflammatory bowel disease

This document describes statistical analyses used to test if inflammatory bowel disease (IBD) affects the somatic mutation burden in the colon. 

```{r cars}
library(ggplot2)
library(cowplot)
library(nlme)
library(lme4)
library(reshape2)


## Read in some data:
########################
crypt_meta <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/sample_info/crypt_metadata_library_bank.txt", h=T)
patient_meta <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/sample_info/patient_metaData.txt", h=T)
crypt_meta <- subset(crypt_meta, WGS_status=="WGS")

mutCount <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/phylogenics/MPBoot/snv/adj_mutCounts/combined_truncBinom_adjusted.txt", h=T)
colnames(mutCount) <- c("crypt_ID", "adj_snv")

mutCount_unadj <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/qc/binomial_filters/combined_Unadj_MutCount.txt", h=T)
mutCount_unadj$crypt_ID <- gsub("_VAF", "", mutCount_unadj$crypt_ID)

## We can now merge all the datasets:
crypt_meta <- merge(crypt_meta, mutCount, by="crypt_ID")
crypt_meta <- merge(crypt_meta, mutCount_unadj, by="crypt_ID")
crypt_meta <- merge(crypt_meta, patient_meta, by="patient_ID")


## Remove patient from control cohort that was massive outlier - high mutation burden due to chemotherapy
crypt_meta <- subset(crypt_meta, patient_ID!="O337")

## Need to clean up the crypt_metadata a bit 
crypt_meta$Disease <- as.factor(as.character(crypt_meta$Disease))
crypt_meta$Disease <- factor(crypt_meta$Disease, levels=c("no_IBD", "CD", "UC"), labels=c("Controls", "CD", "UC"))

test_cohort <- crypt_meta
test_cohort <- subset(test_cohort, location!="Ileum" & location!="Terminal_ileum")

## Set disease duration for never inflamed regions as 0
test_cohort$DiseaseDuration <- as.numeric(as.character(test_cohort$DiseaseDuration))
test_cohort$DiseaseDuration[is.na(test_cohort$DiseaseDuration)] <- 0
test_cohort$DiseaseDuration[test_cohort$type=="Never_Inflamed"] <- 0

## Add 0.5 years to disease duration to avoid 0 in this category for newly diagnosed patients. 
## This is reasonable as IBD symptoms often start months before diagnosis. 
test_cohort$DiseaseDuration[test_cohort$type!="Never_Inflamed"] <- test_cohort$DiseaseDuration[test_cohort$type!="Never_Inflamed"] +0.5

## The IBD dataset has higher spatial resolution than the control dataset. Set the locations as Left, Right and Transverse
test_cohort$location[test_cohort$location %in% c("Ascending", "Cecum", "ICV")] <- "Right"
test_cohort$location[test_cohort$location %in% c("Descending", "Rectum", "Sigmoid")] <- "Left"
test_cohort$location[test_cohort$location %in% c("Hepatic", "Splenic_flexure")] <- "Transverse"
test_cohort$location <- as.factor(as.character(test_cohort$location))

## Exclude technical duplicates etc.
test_cohort <- subset(test_cohort, is.na(excl_criteria))

test_cohort$colour <- "#A72608"
test_cohort$colour[test_cohort$cohort=="control_data"] <- "#BFD7EA"
test_cohort$colour[test_cohort$cohort=="ibd_cohort" & test_cohort$type=="Never_Inflamed"] <- "#FCA311"


head(subset(test_cohort, cohort=="ibd_cohort"))
head(subset(test_cohort, cohort=="control_data"))
dim(subset(test_cohort, cohort=="ibd_cohort"))
dim(subset(test_cohort, cohort=="control_data"))

length(unique(subset(test_cohort, cohort=="ibd_cohort")$patient_ID))
length(unique(subset(test_cohort, cohort=="ibd_cohort" & Disease=="CD")$patient_ID))
length(unique(subset(test_cohort, cohort=="ibd_cohort" & Disease=="UC")$patient_ID))
length(unique(subset(test_cohort, cohort=="control_data")$patient_ID))
```

## Linear mixed effects model for substitution counts

I will use the lme() function of the nlme package to fit a linear mixed effects model to the data. This will account for non-independence of the observations - as crypts are grouped by biopsies, and biopsies are nested within patients. 

The components of the model are:
1. Age and location in the bowel are fixed effects of interest, known from Lee-Six et al. (2019) to affect the somatic mutation burden. 

2. Sequencing coverage and median variant allele fraction are also potentially important contributors but as I carried out an adjustment for these the effect should be modest. 

3. I will allow the slope to vary between biopsy and patient, with biopsy nested within patients. 

4. While the model will allow for random slopes, I fix the intercept at 0. The biological interpretation of this is that there are not somatic mutations at birth (as our filtering strategy will remove most embryonic mutations and these are not of much interest in our study anyway). 

5. I will allow both the within-patient variance and between-variance to vary depending on disease status and formally test if they do so. 


### The null model 

I want to see if there is still some effect of coverage and decide if there should be an interaction effect for Age and location. 

```{r}
# LME model that allows for different variances for each patient and for each corhot (IBD vs controls)

model_null.lme <- lme(fixed = adj_snv ~ Age + location, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_null.lme)
```




### Effect of coverage

Despite us having adjusted the branch lengths of the phylogenic trees by coverage and vaf, some effect remains


```{r}

model_cov <- lme(fixed = adj_snv ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

## Next test to see if Coverage is significant
anova(model_null.lme,model_cov, test=T)$"p-value"[2]
summary(model_cov)

```

We note that the median median VAF is the same in both cohorts while the median coverage of the IBD cohort is slightly higher. This is now shown in Supplementary Figure 2.

```{r}
## Compare the VAF distributions and the coverage between the two different cohorts. 
tmp <- test_cohort
tmp$cohort <- factor(tmp$cohort, labels=c("Controls", "IBD"))
vaf_plot <- ggplot(tmp, aes(x=snvMedianVAF)) + geom_histogram() + facet_grid(cohort~.) + labs(x="Median VAF of isolated crypts")
cov_plot <- ggplot(tmp, aes(x=Coverage)) + geom_histogram() + facet_grid(cohort~.) + labs(x="Median coverage of isolated crypts")

png("/Users/so11/phd/somatic_ibd_p1/manuscript/Re_submission_NG/Supplementary_material/Supplementary_Figure2_vaf_and_cov_histogram.png")
plot_grid(vaf_plot, cov_plot, labels = "AUTO", nrow=2)
dev.off()

tapply(test_cohort$snvMedianVAF, test_cohort$cohort, median)
tapply(test_cohort$Coverage, test_cohort$cohort, median)
```


### Confirm that the model residuals are normally distributed

That the standardized residuals are normally distributed is an assumption of the model. 

```{r}
summary(model_cov)
intervals(model_cov, which = "fixed")

plot(model_cov, resid(., type="p") ~ Age, abline = 0, col=test_cohort$colour, pch=20 )

```

## Including disease duration improves the fit of the linear mixed effects model

I will fit the same model as model_null.lme except I add a fixed effect for Disease Duration, that is time since diagnosis plus 6 months. Never inflamed regions from IBD patients have Disease duration set to 0. 

```{r}

model_dur.lme <- lme(fixed = adj_snv ~ Age + location + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.lme)
intervals(model_dur.lme, which = "fixed")
anova(model_null.lme,model_dur.lme, test=T)$"p-value"[2]

```


The above model does not include Coverage. We can try adding it in and we see that it makes very little difference to the analysis as the effect is small compared with the effect of age and disease duration. 

This is the model used when reporting mutation burden in the main text of the manuscript.

```{r}

model_dur_cov.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_cov.lme)
intervals(model_dur_cov.lme, which = "fixed")
anova(model_cov,model_dur_cov.lme, test=T)$"p-value"[2]


## Supplementary Table 3
sink("/Users/so11/phd/somatic_ibd_p1/manuscript/Re_submission_NG/Supplementary_material/Supplementary_Table3_lmm_summary.txt")
print(summary(model_dur_cov.lme))
sink()

```

## Plot the model: 
```{r}

lme.ints <- intervals(model_dur.lme, which="fixed")$fixed
par(mar=c(5.1,6.1,4.1,2.1))
 plot(NULL, xlim=c(0, (max(test_cohort$Age)+2)), ylim=c(0, (max(test_cohort$adj_snv)+1000)), 
     xlab="Age [Years]", ylab="Substitutions / crypt", yaxs="i", xaxs="i",cex.lab=1, cex.axis=1)


polygon(x = c(0, max(test_cohort$Age), max(test_cohort$Age), 0),
        y = c(lme.ints["(Intercept)", "lower"], 
              lme.ints["(Intercept)", "lower"] + max(test_cohort$Age) * lme.ints["Age", "lower"],
              lme.ints["(Intercept)", "upper"] + max(test_cohort$Age) * lme.ints["Age", "upper"],
              lme.ints["(Intercept)", "upper"]),
        col="#BFD7EA", border=NA)
lines(x = c(0, max(test_cohort$Age)), 
      y = c(lme.ints["(Intercept)", "est."], 
            lme.ints["(Intercept)", "est."] + max(test_cohort$Age) * lme.ints["Age", "est."]),
      lwd=2)

points(x = test_cohort$Age, y = test_cohort$adj_snv, bg=alpha(test_cohort$colour, 0.65), pch=21, cex=1)

legend(x = 5, y = 11700, col = "black", pch=21 ,pt.bg=unique(test_cohort$colour), bty="n", pt.cex = 1,cex=1,
       legend = c("Controls", "IBD - Never inflamed","IBD-with inflammation \n history"))

```




## Restricted analyses

I want to see how this looks like when I only look at specific restricted sub-types of data. We do observe that never-inflamed regions from IBD patients do not have a higher mutation burden than controls. We also observe that there is an effect of disease duration within the IBD cohort. Finally, we can assess the effect of smoking in the IBD cohort. 

```{r}
never_infl_only <- subset(test_cohort, type=="Never_Inflamed")

model_ibd_never_infl1 <- lme(fixed = adj_snv ~ Age + location + cohort + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = never_infl_only, method="ML")

model_ibd_never_infl2 <- lme(fixed = adj_snv ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = never_infl_only, method="ML")

anova(model_ibd_never_infl1,model_ibd_never_infl2, test=T)$"p-value"[2]
summary(model_ibd_never_infl1)


## Just in the diseased samples, is there an effect of disease duration?
ibd_only <- subset(test_cohort, cohort=="ibd_cohort")
model_ibd_only.lme <- lme(fixed = adj_snv ~ Age + location + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~  1), biopsy_ID = pdSymm(form = ~  1)), 
                      weights = varIdent(form= ~ 1 ),
                      data = ibd_only, method="ML")

summary(model_ibd_only.lme)


ibd_smokers <- subset(ibd_only, !is.na(Smoking_years))
ibd_smokers$Disease <- as.factor(as.character(ibd_smokers$Disease))

## Look into the effects of smoking on the mutation burden of IBD patients. 
model_ibd_smoking.lme <- lme(fixed = adj_snv ~ Age + location + DiseaseDuration + Coverage + Smoking_years, 
                      random = list(patient_ID = pdSymm(form = ~  1), biopsy_ID = pdSymm(form = ~  1)), 
                      weights = varIdent(form= ~ 1 ),
                      data = ibd_smokers, method="ML")

model_ibd_only.lme <- lme(fixed = adj_snv ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~  1), biopsy_ID = pdSymm(form = ~  1)), 
                      weights = varIdent(form= ~ 1 ),
                      data = ibd_smokers, method="ML")

anova(model_ibd_only.lme,model_ibd_smoking.lme, test=T)$"p-value"[2]
summary(model_ibd_smoking.lme)
intervals(model_ibd_smoking.lme, which = "fixed")



## Smoking in CD patients vs UC
model_ibd_smoking.uc_cd <- lme(fixed = adj_snv ~ Age + location + DiseaseDuration + Coverage + Smoking_years + Disease, 
                      random = list(patient_ID = pdSymm(form = ~  1), biopsy_ID = pdSymm(form = ~  1)), 
                      weights = varIdent(form= ~ 1 ),
                      data = ibd_smokers, method="ML")

summary(model_ibd_smoking.uc_cd)
anova(model_ibd_smoking.lme,model_ibd_smoking.uc_cd, test=T)$"p-value"[2]

```


## IBD affects the between-patient variance in mutation burden

Next we want to see if the between-patient variance is greater in IBD patients than in the controls. This might be expected as disease severity varies between the patients and patients vary in their response to treatment.  

From the analysis below, we conclude that there is significant difference in the between-patient variance but not in the within-patient variance. 

```{r, eval=F, echo=T}


model_dur.btwbio.het <- lme(fixed = adj_snv ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~cohort-  1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur.btwpt.het <- lme(fixed = adj_snv ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort- 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

anova(model_dur.btwpt.het, model_dur_cov.lme, test=TRUE)$"p-value"[2]
anova(model_dur.btwbio.het, model_dur_cov.lme, test=TRUE)$"p-value"[2]


```



## Repeat the above analyses for indels

```{r}

indel_counts <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/phylogenics/MPBoot/indel/adj_mutCounts/combined_truncBinom_adjusted.txt", h=T)
colnames(indel_counts) <- c("crypt_ID", "adj_indel")

test_cohort <- merge(test_cohort, indel_counts, by="crypt_ID", all.x=T)

model_null.indel <- lme(fixed = adj_indel  ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")


summary(model_null.indel)
intervals(model_null.indel, which = "fixed")

colour.palette.2 <- c("#BFD7EA","#A72608")
names(colour.palette.2) <- c("control_data", "ibd_cohort")
plot(model_null.indel, resid(., type="p") ~ Age, abline = 0, col=colour.palette.2[test_cohort_indel$cohort], pch=20 )


model_dur.indel <- lme(fixed = adj_indel ~ Age + location  + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_cov.indel <- lme(fixed = adj_indel ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_cov.indel)
intervals(model_dur_cov.indel, which = "fixed")
anova(model_null.indel,model_dur_cov.indel, test=T)$"p-value"[2]



## Plot the model
#######################

lme.ints <- intervals(model_dur.indel, which="fixed")$fixed
plot(NULL, xlim=c(0, (max(test_cohort_indel$Age)+2)), ylim=c(0, (max(test_cohort_indel$adj_indel)+100)), 
     xlab="Age [Years]", ylab="InDels / crypt", yaxs="i", xaxs="i")


polygon(x = c(0, max(test_cohort_indel$Age), max(test_cohort_indel$Age), 0),
        y = c(lme.ints["(Intercept)", "lower"], 
              lme.ints["(Intercept)", "lower"] + max(test_cohort_indel$Age) * lme.ints["Age", "lower"],
              lme.ints["(Intercept)", "upper"] + max(test_cohort_indel$Age) * lme.ints["Age", "upper"],
              lme.ints["(Intercept)", "upper"]),
        col="#BFD7EA", border=NA)
lines(x = c(0, max(test_cohort_indel$Age)), 
      y = c(lme.ints["(Intercept)", "est."], 
            lme.ints["(Intercept)", "est."] + max(test_cohort_indel$Age) * lme.ints["Age", "est."]),
      lwd=2)

points(x = test_cohort$Age, y = test_cohort$adj_indel, bg=alpha(test_cohort$colour, 0.65), pch=21, cex=1.5)

legend(x = 0.2, y = 900, col = "black", pch=21 ,pt.bg=unique(test_cohort$colour), bty="n", pt.cex = 1.5,
       legend = c("Controls", "IBD - Never inflamed","IBD-with inflammation \n history"))

```




## LMM for signature exposure

```{r}
SignatureExposure <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/signature_extraction/hdp/snvs/crypt_exposure_mutCounts_post_EM.txt", h=T)

test_cohort <- merge(test_cohort, SignatureExposure, all.x=T)



### Signature 1
##################

model_null.sbs1 <- lme(fixed = SBS1 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBS1)), method="ML")

summary(model_null.sbs1)


model_dur.sbs1 <- lme(fixed = SBS1 ~ Age + location  + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.sbs1)
intervals(model_dur.sbs1, which = "fixed")
anova(model_null.sbs1,model_dur.sbs1, test=T)$"p-value"[2]



### Signature 5
##################
model_null.sbs5 <- lme(fixed = SBS5 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBS5)), method="ML")

summary(model_null.sbs5)


model_dur.sbs5 <- lme(fixed = SBS5 ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBS5)), method="ML")

summary(model_dur.sbs5)
intervals(model_dur.sbs5, which = "fixed")
anova(model_null.sbs5,model_dur.sbs5, test=T)$"p-value"[2]


### Signature 18
####################
model_null.sbs18 <- lme(fixed = SBS18 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBS18)), method="ML")

summary(model_null.sbs18)


model_dur.sbs18 <- lme(fixed = SBS18 ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBS18)), method="ML")

summary(model_dur.sbs18)
intervals(model_dur.sbs18, which = "fixed")
anova(model_null.sbs18,model_dur.sbs18, test=T)$"p-value"[2]



## Signature 32
#######################
model_null.sbs32 <- lme(fixed = SBS32 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_null.sbs32)


model_dur.sbs32 <- lme(fixed = SBS32 ~ Age + location +Coverage+ DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.sbs32)
intervals(model_dur.sbs32, which = "fixed")
anova(model_null.sbs32,model_dur.sbs32, test=T)$"p-value"[2]



### Signature A
####################
model_null.sbsa <- lme(fixed = SBSA ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBSA)), method="ML")

summary(model_null.sbsa)


model_dur.sbsa <- lme(fixed = SBSA ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(SBSA)), method="ML")

summary(model_dur.sbsa)
intervals(model_dur.sbsa, which = "fixed")
anova(model_null.sbsa,model_dur.sbsa, test=T)$"p-value"[2]


## IDA is attributed to the same mutational process as SBSA
#############################################################
model_null.ida <- lme(fixed = IDA ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_null.ida)


model_dur.ida <- lme(fixed = IDA ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.ida)
intervals(model_dur.ida, which = "fixed")
anova(model_null.ida,model_dur.ida, test=T)$"p-value"[2]


## Perhaps it makes more sense to predict IBD status in a logistic regression 
## but the modeling does not allow for unequal variance between the cohorts
## Not used in the paper atm... 
test_cohort$IBD <- 0
test_cohort$IBD[test_cohort$cohort=="ibd_cohort"] <- 1

sbsa_log.null <- glmer(IBD ~ SBSA + location + Coverage + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="binomial", glmerControl(optimizer = "bobyqa", optCtrl = list(maxfun = 100000)))

## ID1
########
model_null.id1 <- lme(fixed = ID1 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(ID1)), method="ML")

summary(model_null.id1)


model_dur.id1<- lme(fixed = ID1 ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.id1)
intervals(model_dur.id1, which = "fixed")
anova(model_null.id1,model_dur.id1, test=T)$"p-value"[2]


## ID2
########
model_null.id2 <- lme(fixed = ID2 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(ID2)), method="ML")

summary(model_null.id2)


model_dur.id2<- lme(fixed = ID2 ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur.id2)
intervals(model_dur.id2, which = "fixed")
anova(model_null.id2,model_dur.id2, test=T)$"p-value"[2]


## IDA
########
model_null.ida <- lme(fixed = IDA ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(IDA)), method="ML")

summary(model_null.ida)


model_dur.ida<- lme(fixed = IDA ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(IDA)), method="ML")

summary(model_dur.ida)
intervals(model_dur.ida, which = "fixed")
anova(model_null.ida,model_dur.ida, test=T)$"p-value"[2]


## ID14
#########

model_null.id14 <- lme(fixed = ID14 ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(IDA)), method="ML")

summary(model_null.id14)


model_dur.id14<- lme(fixed = ID14 ~ Age + location + Coverage + DiseaseDuration, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = subset(test_cohort, !is.na(IDA)), method="ML")

summary(model_dur.id14)
intervals(model_dur.id14, which = "fixed")
anova(model_null.id14,model_dur.id14, test=T)$"p-value"[2]

```


### Differences between CD and UC

I want to find out if there are differences between CD and UC in terms of the mutation burden or eposure to any particular signatures. 

```{r}

ibd_data <- subset(test_cohort, cohort=="ibd_cohort")
ibd_data$Disease <- as.factor(as.character(ibd_data$Disease))


## First look into the mutation burden. Add a fixed effect for Disease. 
## Allow variance to vary between the two diseases. 
model_null.ibd <- lme(fixed = adj_snv ~ Age + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = ibd_data, method="ML")

model_dur.ibd <- lme(fixed = adj_snv ~ Age + DiseaseDuration + Disease, 
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = ibd_data, method="ML")

summary(model_dur.ibd)
intervals(model_dur.ibd, which = "fixed")
anova(model_null.ibd,model_dur.ibd, test=T)$"p-value"[2]


model_null.ibd.indel <- lme(fixed = adj_indel ~ Age +location+ Coverage + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$adj_indel)), method="ML")

model_dur.ibd.indel <- lme(fixed = adj_indel ~ Age +location+ Coverage + DiseaseDuration + Disease, 
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$adj_indel)), method="ML")

summary(model_dur.ibd.indel)
intervals(model_dur.ibd.indel, which = "fixed")
anova(model_null.ibd.indel,model_dur.ibd.indel, test=T)$"p-value"[2]


```

So there is no significant difference between CD and UC in terms of the total mutation burden. 

I next want to test for a difference in specific signature exposure. 

```{r}

model_null.ibd.sbs1 <- lme(fixed = SBS1 ~ Age + location + Coverage + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS1)), method="ML")

model_dur.ibd.sbs1 <- lme(fixed = SBS1 ~ Age + location + Coverage + DiseaseDuration + Disease, 
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS1)), method="ML")

summary(model_dur.ibd.sbs1)
intervals(model_dur.ibd.sbs1, which = "fixed")
anova(model_null.ibd.sbs1,model_dur.ibd.sbs1, test=T)$"p-value"[2]



model_null.ibd.sbs5 <- lme(fixed = SBS5 ~ Age + location + Coverage + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS5)), method="ML")

model_dur.ibd.sbs5 <- lme(fixed = SBS5 ~ Age + location + Coverage + DiseaseDuration*Disease, 
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS5)), method="ML")

summary(model_dur.ibd.sbs5)
intervals(model_dur.ibd.sbs5, which = "fixed")
anova(model_null.ibd.sbs5,model_dur.ibd.sbs5, test=T)$"p-value"[2]



model_null.ibd.sbs18 <- lme(fixed = SBS18 ~ Age + location + Coverage + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS18)), method="ML")

model_dur.ibd.sbs18 <- lme(fixed = SBS18 ~ Age + location + Coverage + DiseaseDuration + Disease, 
                      random = list(patient_ID = pdSymm(form = ~ Disease - 1), biopsy_ID = pdSymm(form = ~ Disease - 1)), 
                      weights = varIdent(form= ~ 1 | Disease),
                      data = subset(ibd_data, !is.na(ibd_data$SBS18)), method="ML")

summary(model_dur.ibd.sbs18)
intervals(model_dur.ibd.sbs18, which = "fixed")
anova(model_null.ibd.sbs18,model_dur.ibd.sbs18, test=T)$"p-value"[2]

```


## Mitochondria 

```{r}
mito <- read.csv("/Users/so11/phd/so11_nfs/somatic_ibd_p1/burden_analyses/colon_mtDNA_coverage.csv", h=T)

ggplot(mito, aes(x=mtDNAgenomes, fill=Study)) + geom_histogram()

x <- merge(test_cohort, mito, by.x="crypt_ID", by.y="Sample")

ggplot(x, aes(x=Age, y=mtDNAgenomes, color=cohort)) + geom_point() + geom_smooth(method="lm")


mito_lme.null <- lme(fixed = mtDNAgenomes ~ Age + location  + Coverage , 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = x, method="ML")

mito_lme.dis <- lme(fixed = mtDNAgenomes ~ Age + location  + Coverage + cohort, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = x, method="ML")

ggplot(x, aes(x=Coverage, y=mtDNAgenomes, color=cohort)) + geom_point()
ggplot(x, aes(x=SBS1, y=mtDNAgenomes, color=cohort)) + geom_point()
ggplot(x, aes(y=mtDNAgenomes, x=cohort)) + geom_boxplot()

ggplot(x, aes(x=lib_conc_.ng.ul., y=mtDNAgenomes)) + geom_point()
ggplot(x, aes(x=lib_conc_.ng.ul., y=mtDNAcoverage, color=cohort)) + geom_point()

## Project 2210 is missing
## Strong negative relationship between Coverage and mtDNAgenomes

```

## Burden of structural variants

We have called deletions, duplications, somatic retrotranspositions and chromosome-level aneuploidies. Below I test each of those for association with disease. 

```{r}

## CNVs
############
cnvs <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/structural_variation/manual_review/Supplementary_Table4_SV_del_and_dup.txt", h=T)

counts <- data.frame(table(cnvs$crypt_ID))
colnames(counts) <- c("crypt_ID", "SV_count")
test_cohort <- merge(test_cohort, counts, by="crypt_ID", all.x=T)

## If the crypt has no calls in the cnvs file then it should have the count of 0
test_cohort$SV_count[is.na(test_cohort$SV_count)] <- 0

sv_model.null <- glmer(SV_count ~ Age + location + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson")
sv_model.dur <- glmer(SV_count ~ Age + location + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson")
anova(sv_model.null,sv_model.dur, test=T)$"Pr(>Chisq)"[2]

summary(sv_model.dur)$coefficients

## Get the 95% confidence intervals for the effect estimate. 
summary(sv_model.dur)$coefficients[5,1] - 1.96*summary(sv_model.dur)$coefficients[5,2]
summary(sv_model.dur)$coefficients[5,1] + 1.96*summary(sv_model.dur)$coefficients[5,2]


ibd_col <- c("#7EB2DD","#F93943",  "#613F75")
p1 <- ggplot(test_cohort, aes(x=Disease, y=SV_count, fill=Disease)) + geom_violin() + 
  geom_jitter(shape=21, position=position_jitter(0.4), aes(colour=Disease), alpha=0.25) +
  scale_fill_manual(values=ibd_col) + scale_colour_manual(values=ibd_col) + 
  labs(x="", y="Number of Copy-number variants") + theme(legend.title = element_blank()) + 
  theme(legend.position = "none")


## Somatic retrotranspositions
##################################

RTs <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/structural_variation/somatic_retrotranspositions/somatic_RTs.txt", h=T)
RT_counts <- data.frame(t(table(RTs$sample)))
RT_counts$Var1 <- NULL
colnames(RT_counts) <- c("crypt_ID", "RT_counts")

test_cohort <- merge(test_cohort, RT_counts, by.x="crypt_ID", by.y="crypt_ID", all.x=T)
test_cohort$RT_counts[is.na(test_cohort$RT_counts)] <- 0

rt_model.null <- glmer(RT_counts ~ Age +location+ Coverage + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson")
rt_model.dur <- glmer(RT_counts ~ Age + location + Coverage +  DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson")
anova(rt_model.null,rt_model.dur, test=T)$"Pr(>Chisq)"[2]

summary(rt_model.dur )$coefficients

## Get the 95% confidence intervals for the effect estimate. 
summary(rt_model.dur )$coefficients[5,1] - 1.96*summary(rt_model.dur)$coefficients[5,2]
summary(rt_model.dur )$coefficients[5,1] + 1.96*summary(rt_model.dur )$coefficients[5,2]

p2 <- ggplot(test_cohort, aes(x=Disease, y=RT_counts, fill=Disease)) + geom_violin() +
  geom_jitter(shape=21, position=position_jitter(0.4), aes(colour=Disease), alpha=0.25) + 
  scale_fill_manual(values=ibd_col) + scale_colour_manual(values=ibd_col) + 
  labs(x="", y="Number of retrotranspositions") + theme(legend.title = element_blank()) + 
  theme(legend.position = "none")


## Aneuploidies 
################

test_cohort$has_chr_event <- NA
test_cohort$has_chr_event[test_cohort$chr_scale_event=="None"] <- 0
test_cohort$has_chr_event[test_cohort$chr_scale_event!="None" & !is.na(test_cohort$chr_scale_event)] <- 1
table(test_cohort$has_chr_event, test_cohort$cohort)
test_cohort$nr_of_chr_events <- 0
test_cohort$nr_of_chr_events[test_cohort$has_chr_event==1] <- 1
test_cohort$nr_of_chr_events[grep(",chr", test_cohort$chr_scale_event)] <- 2

test_cohort$has_IBD <- 0
test_cohort$has_IBD[test_cohort$cohort=="ibd_cohort"] <- 1

aneuploidies_model <- glmer(has_IBD ~ Age +  has_chr_event + (Age - 1|patient_ID/biopsy_ID), data=subset(test_cohort, !(test_cohort$cohort=="ibd_cohort" & test_cohort$type=="Never_Inflamed")), family="binomial")
summary(aneuploidies_model)

aneuploidies_model2_null <- glmer(has_chr_event ~ Age + (Age - 1|patient_ID/biopsy_ID), data=subset(test_cohort, !(test_cohort$cohort=="ibd_cohort" & test_cohort$type=="Never_Inflamed")), family="poisson")

aneuploidies_model2_dur <- glmer(has_chr_event ~ Age + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=subset(test_cohort, !(test_cohort$cohort=="ibd_cohort" & test_cohort$type=="Never_Inflamed")), family="poisson")
summary(aneuploidies_model2_dur)

anova(aneuploidies_model2_null,aneuploidies_model2_dur, test=T)$"Pr(>Chisq)"[2]

crypt_meta$chr_del <- 0
crypt_meta$chr_del[grep("DEL", crypt_meta$chr_scale_event)] <- 1
crypt_meta$chr_dup <- 0
crypt_meta$chr_dup[grep("DUP", crypt_meta$chr_scale_event)] <- 1
crypt_meta$chr_loh <- 0
crypt_meta$chr_loh[grep("LOH", crypt_meta$chr_scale_event)] <- 1
crypt_meta$chr_aneuploidies <- crypt_meta$chr_del + crypt_meta$chr_dup + crypt_meta$chr_loh

frac_ctrl <- c(nrow(crypt_meta[crypt_meta$chr_del==1 & crypt_meta$Disease=="Controls",]) /nrow(crypt_meta[ crypt_meta$Disease=="Controls",]),
               nrow(crypt_meta[crypt_meta$chr_dup==1 & crypt_meta$Disease=="Controls",]) /nrow(crypt_meta[ crypt_meta$Disease=="Controls",]),
               nrow(crypt_meta[crypt_meta$chr_loh==1 & crypt_meta$Disease=="Controls",]) /nrow(crypt_meta[ crypt_meta$Disease=="Controls",]))

frac_CD <- c(nrow(crypt_meta[crypt_meta$chr_del==1 & crypt_meta$Disease=="CD",]) /nrow(crypt_meta[ crypt_meta$Disease=="CD",]),
               nrow(crypt_meta[crypt_meta$chr_dup==1 & crypt_meta$Disease=="CD",]) /nrow(crypt_meta[ crypt_meta$Disease=="CD",]),
               nrow(crypt_meta[crypt_meta$chr_loh==1 & crypt_meta$Disease=="CD",]) /nrow(crypt_meta[ crypt_meta$Disease=="CD",]))

frac_UC <- c(nrow(crypt_meta[crypt_meta$chr_del==1 & crypt_meta$Disease=="UC",]) /nrow(crypt_meta[ crypt_meta$Disease=="UC",]),
             nrow(crypt_meta[crypt_meta$chr_dup==1 & crypt_meta$Disease=="UC",]) /nrow(crypt_meta[ crypt_meta$Disease=="UC",]),
             nrow(crypt_meta[crypt_meta$chr_loh==1 & crypt_meta$Disease=="UC",]) /nrow(crypt_meta[ crypt_meta$Disease=="UC",]))

plot_aneuploidy <- data.frame(aneuploidy=c("Chr Deletion", "Chr Duplication", "Chr LOH"), frac_ctrl,frac_CD, frac_UC)
colnames(plot_aneuploidy) <- c("Aneuploidy", "Controls", "CD", "UC")
plot_aneuploidy_m <- melt(plot_aneuploidy)

p3 <- ggplot(plot_aneuploidy_m, aes(x=variable, y=value, fill=variable)) + geom_bar(position = "dodge", stat="identity") + 
  facet_grid(. ~ Aneuploidy) + theme( axis.title.x = element_blank(),
                                legend.title = element_blank(), axis.ticks.x = element_blank()) + 
  labs(y="Fraction of crypts \n with event") + theme(legend.position = "none") +
  scale_y_continuous(expand = c(0,0), limits=c(0,0.1)) + scale_fill_manual(values = ibd_col)


## Fragile sites
##################

## Enter this region into a genome-browser and inspect manually each genome.
## 16:5050001..8310000 (3.26 Mb) - RBFOX1
## 3:56950001..65100000 (8.15 Mb) - FHIT
## 16:77412001..80082000 (2.67 Mb) WWOX

case_fragile <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/structural_variation/BRASS/p1874_fragile_sites.txt", h=T)
p1728_fragile <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/structural_variation/BRASS/p1728_fragile_sites.txt", h=T)
p1494_fragile <- read.table("/Users/so11/phd/so11_lustre/somatic_ibd_p1/structural_variation/BRASS/p1494_fragile_sites.txt", h=T)
p1494_rbfox <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/structural_variation/BRASS/project1494_16p13.2_deletion_carriers.txt")

p1494_fragile$RBFOX1[p1494_fragile$crypt_ID %in% p1494_rbfox$V1] <- 1

fragile_calls <- rbind(case_fragile, p1728_fragile, p1494_fragile)
fragile_calls$patient_ID <- NULL

test_cohort <- merge(test_cohort, fragile_calls, by="crypt_ID", all.x=T)
table(test_cohort$RBFOX1, test_cohort$cohort)
table(test_cohort$FHIT, test_cohort$cohort)
table(test_cohort$WWOX, test_cohort$cohort)

rbfox1_model.null <- glmer(RBFOX1 ~ Age + location + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
rbfox1_model.dur <- glmer(RBFOX1 ~ Age + location + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
anova(rbfox1_model.null,rbfox1_model.dur, test=T)$"Pr(>Chisq)"[2]

fhit_model.null <- glmer(FHIT ~ Age + location + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
fhit_model.dur <- glmer(FHIT ~ Age + location + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
anova(fhit_model.null,fhit_model.dur, test=T)$"Pr(>Chisq)"[2]

wwox_model.null <- glmer(WWOX ~ Age + location + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
wwox_model.dur <- glmer(WWOX ~ Age + location + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
anova(wwox_model.null,wwox_model.dur, test=T)$"Pr(>Chisq)"[2]

test_cohort$all_fragile <- test_cohort$RBFOX1 + test_cohort$WWOX + test_cohort$FHIT
all_fragile_model.null <- glmer(all_fragile ~ Age + location + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
all_fragile_model.dur <- glmer(all_fragile ~ Age + location + DiseaseDuration + (Age - 1|patient_ID/biopsy_ID), data=test_cohort, family="poisson", nAGQ = 0)
anova(all_fragile_model.null,all_fragile_model.dur, test=T)$"Pr(>Chisq)"[2]


frags_data <- frags_data[is.na(frags_data$excl_criteria) & frags_data$patient_ID.x!="O337" & 
                            !is.na(frags_data$RBFOX1), c("crypt_ID", "Disease", "type", "RBFOX1", "FHIT", "WWOX")]

## Want to show fragile sites in fractions of crypts that have inflammation history
frags_data <- subset(frags_data, !(frags_data$Disease!="Controls" & type=="Never_Inflamed"))


## There are some rare duplications at the fragile sites. We are going to exclude
## those from the analysis as their interpretation is not clear. 
frags_data$RBFOX1[frags_data$RBFOX1>2] <- NA 
frags_data$RBFOX1[frags_data$WWOX>2] <- NA
frags_data$RBFOX1[frags_data$FHIT>2] <- NA

frac_ctrl <- c(nrow(frags_data[frags_data$FHIT!=2 & frags_data$Disease=="Controls",]) /nrow(frags_data[ frags_data$Disease=="Controls",]),
               nrow(frags_data[frags_data$RBFOX1!=2 & frags_data$Disease=="Controls",]) /nrow(frags_data[ frags_data$Disease=="Controls",]),
               nrow(frags_data[frags_data$WWOX!=2 & frags_data$Disease=="Controls",]) /nrow(frags_data[ frags_data$Disease=="Controls",]))

frac_CD <- c(nrow(frags_data[frags_data$FHIT!=2 & frags_data$Disease=="CD",]) /nrow(frags_data[ frags_data$Disease=="CD",]),
               nrow(frags_data[frags_data$RBFOX1!=2 & frags_data$Disease=="CD",]) /nrow(frags_data[ frags_data$Disease=="CD",]),
               nrow(frags_data[frags_data$WWOX!=2 & frags_data$Disease=="CD",]) /nrow(frags_data[ frags_data$Disease=="CD",]))

frac_UC <- c(nrow(frags_data[frags_data$FHIT!=2 & frags_data$Disease=="UC",]) /nrow(frags_data[ frags_data$Disease=="UC",]),
               nrow(frags_data[frags_data$RBFOX1!=2 & frags_data$Disease=="UC",]) /nrow(frags_data[ frags_data$Disease=="UC",]),
               nrow(frags_data[frags_data$WWOX!=2 & frags_data$Disease=="UC",]) /nrow(frags_data[ frags_data$Disease=="UC",]))


plot_data <- data.frame(Locus=c("chr3p14.2 (FHIT)", "chr16p13.2 (RBFOX1)", "chr16q23.1 (WWOX)"), frac_ctrl,frac_CD, frac_UC)
colnames(plot_data) <- c("Locus", "Controls", "CD", "UC")
plot_data_m <- melt(plot_data)

p4 <- ggplot(plot_data_m, aes(x=variable, y=value, fill=variable)) + geom_bar(position = "dodge", stat="identity") + 
  facet_grid(. ~ Locus) + theme(axis.title.x = element_blank(),
                                legend.title = element_blank(), axis.ticks.x = element_blank()) + 
  labs(y="Fraction of crypts \ncarrying a deletion") + theme(legend.position = "none") + 
  scale_y_continuous(expand = c(0,0), limits=c(0,0.35)) + scale_fill_manual(values =ibd_col)



## Make figure 3
#plot_grid(plot_grid(p1,p2, labels=c("a", "b")), p3, p4, nrow=3, rel_heights = c(2,1,1), labels=c("","c", "d"))
plot_grid(plot_grid(p1,p2, labels=c("A", "B")), p3, nrow=2, rel_heights = c(2,1), labels=c("","C"))


```



## Predict IBD mutation burden

This is what is shown in Figure 1b. 

I want to train a model on the control data and use it to predict mutation burden in the IBD data. 
We have to abandon the modeling of different variances in the cohort and predictions can only be made at the cohort level (if Coverage isn't included, the numbers for crypts from the same patient are identical). There is no way to estimate random effects so
"The predictions at level i are obtained by adding together the contributions from the estimated fixed effects and the estimated random effects at levels less or equal to i and evaluating the model function at the resulting estimated parameters. If group values not included in the original grouping factors are present in newdata, the corresponding predictions will be set to NA for levels greater or equal to the level at which the unknown groups occur." (From the predict.nlme() documentation)

This obviously means that the model is not exactly the same as the one used to report effects in the main text, but I only use it to graphically convey the message of disease duration being associated with increased mutation burden. 

```{r}
# https://rdrr.io/cran/nlme/man/predict.nlme.html

ctrl <- subset(test_cohort, cohort=="control_data")
ibd <-  subset(test_cohort, cohort=="ibd_cohort" & type!="Never_Inflamed")

model_null.ctrl <- lme(fixed = adj_snv ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ctrl, method="ML")

ibd$predict <- predict(model_null.ctrl,ibd,level=0 )
ibd$delta <- ibd$adj_snv - ibd$predict

model_predict.ibd <- lme(fixed = delta ~ DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

summary(model_predict.ibd)

## Repeat the above for indels:
##################################


model_null.ctrl_id <- lme(fixed = adj_indel ~ Age + location + Coverage, 
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ctrl, method="ML")

ibd$predict_id <- predict(model_null.ctrl_id,ibd,level=0 )
ibd$delta_id <- ibd$adj_indel - ibd$predict_id

model_predict_id.ibd <- lme(fixed = delta_id ~ DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

summary(model_predict_id.ibd)
```

## Check if drivers are associated with mutation burden and signatures


```{r}
driver_counts <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/burden_analyses/driver_counts.txt", h=T)
test_cohort <- merge(test_cohort, driver_counts, by.x="crypt_ID", by.y="crypt_ID", all.x=T)


model_dur_driver.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_driver.lme)
intervals(model_dur_driver.lme, which = "fixed")
anova(model_dur_driver.lme,model_dur_cov.lme, test=T)$"p-value"[2]

## See if driver status also influences signature exposure:
model_dur_driver.sbs1 <- lme(fixed = SBS1 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.sbs5 <- lme(fixed = SBS5 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.sbs18 <- lme(fixed = SBS18 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.sbsa <- lme(fixed = SBSA ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.sbs32 <- lme(fixed = SBS32 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")



model_dur_driver.id1 <- lme(fixed = ID1 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.id2 <- lme(fixed = ID2 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.ida <- lme(fixed = IDA ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

model_dur_driver.id14 <- lme(fixed = ID14 ~ Age + location +Coverage+ DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")


anova(model_dur.sbs1,model_dur_driver.sbs1, test=T)$"p-value"[2]
anova(model_dur.sbs5,model_dur_driver.sbs5, test=T)$"p-value"[2]
anova(model_dur.sbs18,model_dur_driver.sbs18, test=T)$"p-value"[2]

anova(model_dur.sbsa,model_dur_driver.sbsa, test=T)$"p-value"[2]
anova(model_dur.sbs32,model_dur_driver.sbs32, test=T)$"p-value"[2]

anova(model_dur.id1,model_dur_driver.id1, test=T)$"p-value"[2]
anova(model_dur.id2,model_dur_driver.id2, test=T)$"p-value"[2]
anova(model_dur.ida,model_dur_driver.ida, test=T)$"p-value"[2]
anova(model_dur.id14,model_dur_driver.id14, test=T)$"p-value"[2]

summary(model_dur_driver.ida)




model_null.ibd.sbs32_null <- lme(fixed = SBS32 ~ Age + location + Coverage + DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      weights = varIdent(form= ~ 1),
                      data = ibd, method="ML")

model_dur.ibd.sbs32 <- lme(fixed = SBS32 ~ Age + location + Coverage + DiseaseDuration + cancer_driver, 
                      random = list(patient_ID = pdSymm(form = ~  1), biopsy_ID = pdSymm(form = ~  1)), 
                      weights = varIdent(form= ~ 1 ),
                      data = ibd, method="ML")



model_dur_RTs.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration + RT_counts, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_RTs.lme)
intervals(model_dur_RTs.lme, which = "fixed")
anova(model_dur_RTs.lme,model_dur_cov.lme, test=T)$"p-value"[2]


model_dur_cnvs.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration + SV_count, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_cnvs.lme)
intervals(model_dur_cnvs.lme, which = "fixed")
anova(model_dur_cnvs.lme,model_dur_cov.lme, test=T)$"p-value"[2]


model_dur_aneu.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration + nr_of_chr_events, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_aneu.lme)
intervals(model_dur_aneu.lme, which = "fixed")
anova(model_dur_aneu.lme,model_dur_cov.lme, test=T)$"p-value"[2]

model_dur_fragile.lme <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration +all_fragile, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_dur_fragile.lme)
intervals(model_dur_fragile.lme, which = "fixed")
anova(model_dur_fragile.lme,model_dur_cov.lme, test=T)$"p-value"[2]

## Since fragile site deletions, RTs, CNVs and aneuploidies may all be correlated, it's probably
## best to estimate the effects of all in the same model: 

model_events.ibd <- lme(fixed = adj_snv ~ Age + location +Coverage+ DiseaseDuration + SV_count +RT_counts + nr_of_chr_events, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_events.ibd)
intervals(model_events.ibd, which = "fixed")
summary(model_events.ibd)$tTable




model_drivers.ibd.null <- lme(fixed = delta_id ~ DiseaseDuration,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

model_drivers.ibd <- lme(fixed = delta_id ~ DiseaseDuration + cancer_driver,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

summary(model_drivers.ibd)
anova(model_drivers.ibd,model_drivers.ibd.null, test=T)$"p-value"[2]



model_RTs.ibd <- lme(fixed = delta_id ~ DiseaseDuration + RT_counts,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

summary(model_RTs.ibd)
anova(model_RTs.ibd,model_drivers.ibd.null, test=T)$"p-value"[2]


model_cnvs.ibd <- lme(fixed = delta_id ~ DiseaseDuration + SV_count,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

anova(model_cnvs.ibd,model_drivers.ibd.null, test=T)$"p-value"[2]

model_aneu.ibd <- lme(fixed = delta ~ DiseaseDuration + nr_of_chr_events + SV_count + RT_counts + cancer_driver.x + ibd_driver.x,
                      random = list(patient_ID = pdSymm(form = ~ 1), biopsy_ID = pdSymm(form = ~ 1)), 
                      data = ibd, method="ML")

anova(model_aneu.ibd,model_drivers.ibd.null, test=T)$"p-value"[2]



model_events.ibd_id <- lme(fixed = adj_indel ~ Age + location +Coverage+ DiseaseDuration + SV_count +RT_counts + nr_of_chr_events, 
                      random = list(patient_ID = pdSymm(form = ~ cohort - 1), biopsy_ID = pdSymm(form = ~ cohort - 1)), 
                      weights = varIdent(form= ~ 1 | cohort),
                      data = test_cohort, method="ML")

summary(model_events.ibd_id)
intervals(model_events.ibd_id, which = "fixed")
summary(model_events.ibd_id)$tTable
```


## Write out the data to a supplementary table:

```{r}
Suppl_table_2 <- test_cohort[, c("crypt_ID", "patient_ID","Sex","Age","cohort", "Disease", "DiseaseDuration", 
                                 "Purine_years","Smoking_years","biopsy_ID","location", "type","Coverage",
                                 "snvMedianVAF", "indelMedianVAF","adj_snv", "adj_indel","SV_count","RT_counts","chr_scale_event",
                                 "Unassigned_SBS", "SBS1", "SBSA", "SBS18","SBS5", "SBS32",
                                 "SBSB","SBSC", "SBS17b", "SBS35", "SBS13", "SBS2", "SBS17a",
                                 "Unassigned_ID", "ID1", "IDA", "ID2", "IDB", "ID14")]


## Update some of the names to resolve sample swaps etc.
new_names <- read.table("/Users/so11/phd/so11_nfs/somatic_ibd_p1/sample_info/sample_swap_rename.txt", h=T, stringsAsFactors = F)

Suppl_table_2$crypt_ID <- as.character(Suppl_table_2$crypt_ID)
for(i in 1:nrow(Suppl_table_2)) {
  if(Suppl_table_2$crypt_ID[i] %in% new_names$original_name) {
    newName <- new_names$new_name[which(new_names$original_name==Suppl_table_2$crypt_ID[i])]
    Suppl_table_2$crypt_ID[i] <- newName
  }
}

write.table(Suppl_table_2, "/Users/so11/phd/somatic_ibd_p1/manuscript/Re_submission_NG/Supplementary_material/Supplementary_Table2_crypt_metaData.txt", row.names=F, quote=F, sep="\t")
```







